imgee/views/files.py

```diff
@@ -0,0 +1,133 @@
+# -*- coding: utf-8 -*-
+from flask import (flash, g, jsonify, redirect, render_template, request, url_for)
+from sqlalchemy import and_
+
+from coaster.views import load_model, load_models
+from imgee import app, lastuser
+from imgee.forms import (
+    UploadImageForm, UpdateTitle, EditTitleForm,
+    AddLabelForm, DeleteImageForm
+)
+from imgee.models import StoredFile, db, Profile
+from imgee.storage import delete, save_file
+from imgee.utils import get_media_domain
+from .index import get_prev_next_images
+import imgee.utils as utils
+
+
+def _redirect_url_frm_upload(profile_name):
+    # if the referrer is from 'pop_up_gallery' redirect back to referrer.
+    referrer = request.referrer or ''
+    if url_for('pop_up_gallery', profile=profile_name) in referrer:
+        url = request.referrer
+    else:
+        url = url_for('profile_view', profile=profile_name)
+    return url
+
+
+@app.route('/<profile>/new', methods=['GET', 'POST'])
+@lastuser.requires_login
+@load_model(Profile, {'name': 'profile'}, 'profile',
+    permission=['new-file', 'siteadmin'])
+def upload_file(profile):
+    upload_form = UploadImageForm()
+    if upload_form.validate_on_submit():
+        file_ = request.files['file']
+        title, stored_file = save_file(file_, profile=profile)
+        flash('"%s" uploaded successfully.' % title)
+        return redirect(_redirect_url_frm_upload(profile.name))
+    return render_template('form.html', form=upload_form, profile=profile)
+
+
+@app.route('/<profile>/new.json', methods=['POST'])
+@lastuser.requires_login
+@load_model(Profile, {'name': 'profile'}, 'profile',
+    permission=['new-file', 'siteadmin'])
+def upload_file_json(profile):
+    upload_form = UploadImageForm()
+    if upload_form.validate_on_submit():
+        file_ = request.files['file']
+        title, stored_file = save_file(file_, profile=profile)
+        update_form = UpdateTitle()
+        update_form.title.data = stored_file.title
+        form = render_template('edit_title_form.html', form=update_form, formid='edit_title_' + stored_file.name)
+        return jsonify(
+            status=True, message="%s uploaded successfully" % title, form=form,
+            update_url=url_for('update_title_json', profile=profile.name, file=stored_file.name),
+            image_data=stored_file.dict_data())
+    else:
+        response = jsonify(status=False, message=' '.join(upload_form.errors.values()))
+        response.status_code = 403
+        return response
+
+
+@app.route('/<profile>/edit_title', methods=['POST'])
+@lastuser.requires_login
+@load_model(Profile, {'name': 'profile'}, 'profile',
+    permission=['edit', 'siteadmin'])
+def edit_title(profile):
+    form = EditTitleForm()
+    if form.validate_on_submit():
+        file_name = request.form.get('file_name')
+        q = and_(Profile.userid.in_(g.user.user_organizations_owned_ids()), StoredFile.name == file_name)
+        f = StoredFile.query.filter(q).first_or_404()
```

Comment:
if the `edit` permission is granted correctly in the `Profile` model, the `abort` check may be unnecessary. Confirm that.